packages/flutter/lib/src/foundation/serialization.dart

```diff
@@ -16,85 +16,101 @@ class WriteBuffer {
   void putUint32(int value, {Endian? endian}) {
+    assert(!_isDone);
     _eightBytes.setUint32(0, value, endian ?? Endian.host);
-    _buffer!.addAll(_eightBytesAsList, 0, 4);
+    _buffer.addAll(_eightBytesAsList, 0, 4);
   }
 
   /// Write an Int32 into the buffer.
   void putInt32(int value, {Endian? endian}) {
+    assert(!_isDone);
     _eightBytes.setInt32(0, value, endian ?? Endian.host);
-    _buffer!.addAll(_eightBytesAsList, 0, 4);
+    _buffer.addAll(_eightBytesAsList, 0, 4);
   }
 
   /// Write an Int64 into the buffer.
   void putInt64(int value, {Endian? endian}) {
+    assert(!_isDone);
     _eightBytes.setInt64(0, value, endian ?? Endian.host);
-    _buffer!.addAll(_eightBytesAsList, 0, 8);
+    _buffer.addAll(_eightBytesAsList, 0, 8);
   }
 
   /// Write an Float64 into the buffer.
   void putFloat64(double value, {Endian? endian}) {
+    assert(!_isDone);
     _alignTo(8);
     _eightBytes.setFloat64(0, value, endian ?? Endian.host);
-    _buffer!.addAll(_eightBytesAsList);
+    _buffer.addAll(_eightBytesAsList);
   }
 
   /// Write all the values from a [Uint8List] into the buffer.
   void putUint8List(Uint8List list) {
-    _buffer!.addAll(list);
+    assert(!_isDone);
+    _buffer.addAll(list);
   }
 
   /// Write all the values from an [Int32List] into the buffer.
   void putInt32List(Int32List list) {
+    assert(!_isDone);
     _alignTo(4);
-    _buffer!.addAll(list.buffer.asUint8List(list.offsetInBytes, 4 * list.length));
+    _buffer.addAll(list.buffer.asUint8List(list.offsetInBytes, 4 * list.length));
   }
 
   /// Write all the values from an [Int64List] into the buffer.
   void putInt64List(Int64List list) {
+    assert(!_isDone);
     _alignTo(8);
-    _buffer!.addAll(list.buffer.asUint8List(list.offsetInBytes, 8 * list.length));
+    _buffer.addAll(list.buffer.asUint8List(list.offsetInBytes, 8 * list.length));
   }
 
   /// Write all the values from a [Float64List] into the buffer.
   void putFloat64List(Float64List list) {
+    assert(!_isDone);
     _alignTo(8);
-    _buffer!.addAll(list.buffer.asUint8List(list.offsetInBytes, 8 * list.length));
+    _buffer.addAll(list.buffer.asUint8List(list.offsetInBytes, 8 * list.length));
   }
 
   void _alignTo(int alignment) {
-    final int mod = _buffer!.length % alignment;
+    assert(!_isDone);
+    final int mod = _buffer.length % alignment;
     if (mod != 0) {
-      for (int i = 0; i < alignment - mod; i++)
-        _buffer!.add(0);
+      _eightBytes.setInt64(0, 0);
+      _buffer.addAll(_eightBytesAsList, 0, alignment - mod);
     }
   }
 
   /// Finalize and return the written [ByteData].
   ByteData done() {
-    final ByteData result = _buffer!.buffer.asByteData(0, _buffer!.lengthInBytes);
-    _buffer = null;
+    if (_isDone) {
+      throw StateError('done() must not be called more than once on the same $runtimeType.');
+    }
+    final ByteData result = _buffer.buffer.asByteData(0, _buffer.lengthInBytes);
+    _isDone = true;
```

Comment:
You should still null out buffer here, otherwise your changing the ownership semantics and GC won't be able to clean up code it used to.

Reply:
(I realize this means having nullchecks on every access, but it's better to take that slight runtime hit than to potentially retain a large chunk of memory unexpectedly)

Reply:
That would only happen if someone kept ahold of a WriteBuffer after they've called `done()`.  There is literally nothing you can do with a WriteBuffer after you've called `done()` so that doesn't make sense.

Reply:
Someone could have existing code where their WriteBuffer is not nullable, and they don't even realize they're holding on to it but they are - but since we null out the underlying buffer, it's ok, whereas now it is not.

Reply:
I think that's an unlikely hypothetical and even still I think there is an argument that keeping a Buffer around should be expected to occupy memory.  Here is our actual usage: https://github.com/flutter/flutter/blob/f6f59c58bcd6ee446ac95c081be1d42e62b58065/packages/flutter/lib/src/services/message_codecs.dart#L313

Reply:
We ran into this when refactoring `Widget` and `State` and `Element` in the framework - things people shouldn't hold on to, but it became super easy to accidentally hold on to because of leaking a BuildContext into a closure that wasn't obviously retained, but ended up being retained because of other framework code, etc.

Reply:
Right, the difference here is that any method called on that object after `done()` causes an exception.  I'll bet that wasn't the case you were running into, because it's nonsensical.  Everywhere we use the class it is a temporary variable which makes the most sense how others will use it, given its semantics.  I don't believe we have a policy blocking changes that have memory ramifications on hypothetical client code that runs contrary to the intent of the class.  We do have policy surrounding syntax breaking changes and changes that affect negatively affects benchmarks, which this won't.

The performance of these methods are key to Flutter performance since they are used for channels, it's not a trivial amount of savings given the bandwidth given to these functions.  My work is building directly on top of them and will increase even further the bandwidth.

Reply:
It's very, very similar to the case we were running into - the problem wasn't that users were calling methods on the object incorrectly, it's that they don't realize it can't be GC'd because it got retained by some closure they stuffed it into that is still reachable.

Reply:
Maybe we can set up a GVC on this on Monday - I'm not suggesting some policy on blocking changes for hypothetical memory impacts, but changing the ownership semantics of this should be considered breaking IMO. The main problem is that it's breaking in a way that's really hard to test for, and a way that our clients have a really hard time figuring out on their own right now.

Reply:
It would help if you have benchmark numbers showing the difference adding the nullchecks in create. If it makes that much of a difference, we could also consider creating a new class that doesn't null things out at the end that's for private use in the framework.

Reply:
Yea, I was surprised we didn't have benchmarks for this already.  Logically we know this is a hot spot since all platform channels go through it multiple times per message (both ways, sending and receiving a response).  One problem with the benchmarks would be that since null-safety is newish they might not be doing all the optimizations they can do.  They might very well still be doing a null check when calling methods, but by expressing our code in a way that is a subset of operations we can reasonably say at worst it will have the same performance and has the opportunity to only improve with compiler improvements.  It would be a huge problem for Dart if it was actually slower.

Reply:
I mainly would just expect that the null check is only a few instructions and doesn't add significantly to runtime for the number of ops we end up doing.

Reply:
One more thing I realized over the weekend.  This code doesn't change the memory profile as much as you think since the ByteData is a view on the same buffer.  So, when we null out _buffer, it isn't actually cleaned up until the returned ByteData is garbage collected.

Reply:
@dnfield Hixie had the good idea just to make `_buffer` not final and replacing it with a zero buffer.  This way we get the performance we want and the same memory characteristics.

Reply:
As long as that doesn't zero out the underlying buffer that works for me.

Reply:
Zero out the underlying buffer that we return to the caller I mean

Reply:
It doesn't.  Can you remove your changes requested, please.